src/public/modules/forms/services/form-api.client.factory.js

```diff
@@ -25,129 +29,170 @@ const extractFormId = (path) => {
    */
-  const getInterceptor = (redirectOnError, errorTargetState) => {
-    const interceptor = {
-      request: (config) => {
-        if (get(config, 'data.form')) {
-          FormFields.removeMyInfoFromForm(config.data.form)
-        }
-        return config
-      },
-      response: (response) => {
-        // The backend returns different shapes for different request types. For GET
-        // requests, it returns a $resource instance with a form attribute containing
-        // all the form data; for PUT requests, it returns a $resource instance with
-        // all the form data at the top level. We need to ensure that the postprocessing
-        // is done in both cases.
-        if (get(response, 'resource.form.form_fields')) {
-          FormFields.injectMyInfoIntoForm(response.resource.form)
-          // Convert plain form object to smart Form instance
-          response.resource.form = new Form(response.resource.form)
-        } else if (get(response, 'resource.form_fields')) {
-          FormFields.injectMyInfoIntoForm(response.resource)
-          response.resource = new Form(response.resource)
-        }
-        return response.resource
-      },
+  const injectMyInfo = (data) => {
+    // The backend returns different shapes for different request types. For GET
+    // requests, it returns an object with a form attribute containing
+    // all the form data; for PUT requests, it returns an object with
+    // all the form data at the top level. We need to ensure that the postprocessing
+    // is done in both cases.
+    if (get(data, 'form.form_fields')) {
+      FormFields.injectMyInfoIntoForm(data.form)
+      // Convert plain form object to smart Form instance
+      data.form = new Form(data.form)
+    } else if (get(data, 'form_fields')) {
+      FormFields.injectMyInfoIntoForm(data)
+      data = new Form(data)
     }
+    return data
+  }
+
+  /**
+   * Handles error returned from API calls. In charge of redirecting users using FormErrorService
+   * if redirectOnError is true.
+   * @param {Error} err Error returned from API calls
+   * @param {{redirectOnError: boolean, errorTargetState: string }} errorParams redirectOnError specifies
+   * whether to use FormErrorService to redirect to errorTargetState when an error is encountered.
+   * If redirectOnError is true, this is the redirect state which will be passed to FormErrorService
+   */
+  const handleError = (err, errorParams) => {
+    const { redirectOnError, errorTargetState } = errorParams
     if (redirectOnError) {
-      interceptor.responseError = (response) => {
-        return FormErrorService.redirect({
-          response,
-          targetState: errorTargetState,
-          targetFormId: extractFormId(get(response, 'config.url')),
-        })
-      }
+      FormErrorService.redirect({
+        response: err.response,
+        targetState: errorTargetState,
+        targetFormId: extractFormId(get(err.response, 'config.url')),
```

Comment:
I actually think the current implementation of `extractFormId` is wrong because we don't always have the form ID at the start of the path, but that can be fixed separately